rules.d/90-iocost.rules

```diff
@@ -0,0 +1,20 @@
+#  SPDX-License-Identifier: LGPL-2.1-or-later
+#
+#  This file is part of systemd.
+#
+#  systemd is free software; you can redistribute it and/or modify it
+#  under the terms of the GNU Lesser General Public License as published by
+#  the Free Software Foundation; either version 2.1 of the License, or
+#  (at your option) any later version.
+
+SUBSYSTEM!="block", GOTO="iocost_end"
+
+ACTION=="remove", GOTO="iocost_end"
+
+IMPORT{builtin}="hwdb 'block::name:$env{ID_MODEL}:fwrev:$env{ID_REVISION}:'"
```

Comment:
See https://github.com/systemd/systemd/pull/23325#discussion_r1147641930.
But I agree that the reasoning needs to be explained.

Reply:
hmm? what is the reasoning? that there are already hwdb records formatted with a key like this out there? is that what you are saying?

Reply:
I didn't touch hwdb file because it's a bit non-trivial. hwdb files are generated from iocost-benchmarks processing of JSON based benchmark result and I would need to make modifications there. Also, I wasn't sure if there is no side-effect of changing hwdb schema. I will confirm this with @kov

Reply:
On rethinking the pattern again it seems to me that having a * is required here:
Explanation:
`IMPORT{builtin}="hwdb 'block:*:name:$env{ID_MODEL}:fwrev:$env{ID_REVISION}:'"`
with above,  * should to match to anything( either * or string).  
So it would match with the following in hwdb file.
```
block:*:name:WDC CL SN730 SDBQNTY-512G-2020:fwrev:*:
OR
block:/dev/sdb:name:WDC CL SN730 SDBQNTY-512G-2020:fwrev:*:  // /dev/sdb after block:
```
But If *  is removed from udev rules then it won't match block:**/dev/sdb**:name:WDC CL SN730 SDBQNTY-512G-2020:fwrev:**4567**: which can be part of hwdb later on.

I'm not udev expert, so I might be wrong here so suggestions are welcome.

Reply:
I tried to explain this in https://github.com/systemd/systemd/pull/23325#discussion_r1147641930. Let me try again more directly:
you're confusing the left and right sides of a match operator (the string and the pattern).
In the udev rule we generate a **literal string**, and then in the hwdb we have **match patterns**.

For example, in the hwdb we have the following pattern:
```python
evdev:input:b0003v05FEp1010*
 XKB_FIXED_LAYOUT=us
 XKB_FIXED_VARIANT=
```
We can check the matches against this pattern:
```console
$ systemd-hwdb query 'evdev:input:b0003v05FEp1010:ABCDEFGH'
XKB_FIXED_LAYOUT=us
XKB_FIXED_VARIANT=
```
A somewhat confusing situation occurs when the pattern is matched against itself:
```console
$ systemd-hwdb query 'evdev:input:b0003v05FEp1010*'
XKB_FIXED_LAYOUT=us
XKB_FIXED_VARIANT=
```
This matches, because the literal string "*" is matched by the wildcard "*" in the pattern.

If we reverse the operands, the match will not occur (note the change of 1010 → 2020):
```python
evdev:input:b0003v05FEp2020:ABCDEFGH
 XKB_FIXED_LAYOUT=xx
 XKB_FIXED_VARIANT=
```
```console
$ sudo systemd-hwdb update
$ systemd-hwdb query 'evdev:input:b0003v05FEp2020:ABCDEFGH'
XKB_FIXED_LAYOUT=xx
XKB_FIXED_VARIANT=
$ systemd-hwdb query 'evdev:input:b0003v05FEp2020*'
(nothing)
```
The last query does not match, because the literal string "*" is not matched by the ":ABCDEFGH" in the pattern.

In general, it makes no sense to have `*` in the rules file. It is useful only in the hwdb file.

Reply:
Thanks for clarifying and I agree * should completely be gone from udev rules.
So I am digging more about the best pattern in udev rules and on hwdb file. 

The issue now surfaced is as:

Right now we have a match based on `name:ID_MODEL` and `fwrev:ID_REVISION` and if it resulted in empty then we use to fallback by matching with the only `name:ID_MODEL` and any value of fwrev (with `fwrev:*`) , but since * has no meaning in udev literal string, it seems there is no way to do fallback mechanism.

e.g hwdb: `block:*:name:WDC123:fwrev:3458: `will not match with `block::name:WDC123:fwrev:1234:` in first trigger but will match in fallback trigger( which take any value of fwrev).

I hope to get better rules and hwdb format here.

Reply:
On diving deeper, it seems having perfect matching rules will need quite an effort since it requires changing of hwdb pattern syntax. hwdb matching pattern is generated by resctl-bench  and thus would need to have a broader agreement.

Thus for now, only removing  * from udev rules(as suggested in https://github.com/systemd/systemd/pull/23325#discussion_r1147641930) and leaving hwdb unchanged would be good enough in the present scenario.

Reply:
On diving deeper, it seems having perfect matching rules will need quite an effort since it requires changing of hwdb pattern syntax. hwdb matching pattern is generated by resctl-bench  and thus would need to have a broader agreement.

Thus for now, only removing  * from udev rules(as suggested in https://github.com/systemd/systemd/pull/23325#discussion_r1147641930) and leaving hwdb unchanged would be good enough in the present scenario.

Reply:
Does resctl-bench have any other users? Updating it now to DTRT would seem fairly trivial.

Reply:
OK, let's ignore the question of the doubled colon. It'd be nicer to get rid of this, but it's not terribly important. 

The improt should be guarded by `ENV{ID_MODEL}!=""`, to avoid a call that cannot succeed.